pkg/messages.go

```diff
@@ -0,0 +1,182 @@
+package giga
+
+/*
+The message subsystem exists to allow event-based access to
+the various parts of GigaWallet's processes, for integration purposes.
+
+A simple internal 'message bus' is passed around internally as a
+singleton, with an internal goroutine and a 'send' method for sending
+'messages'.
+
+outbound destinations are created in config, which result in these
+messages being routed to various external services, ie: MQTT, AMQP,
+HTTP callbacks, log-files, etc. These are managed by MessageSubscribers:
+
+MessageSubscribers are registered with the bus and are subscribed via
+their own channels along with a list of MessageTypes they want to subscrive
+to.
+*/
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+)
+
+type MessageType string
+
+// These consts are used to pub and sub to messages
+const (
+	MSG_ALL MessageType = "ALL" // Do not use for sending
+	MSG_SYS MessageType = "SYS" // System messages
+	MSG_NET MessageType = "NET" // Network Events
+	MSG_ACC MessageType = "ACC" // Account Events
+	MSG_INV MessageType = "INV" // Innvoice Events
+)
+
+// slice of all msg types for config funcs lookup
+var MSG_TYPES []MessageType = []MessageType{MSG_ALL,
+	MSG_SYS, MSG_NET, MSG_ACC, MSG_INV}
+
+// MessageSubscribers are things that subscribe to the bus and handle
+// messages, ie: MQTT, AMQP, http callbacks etc.
+type MessageSubscriber interface {
+	GetChan() chan Message
+}
+
+// Created by the bus, wraps message sent with Send
+type Message struct {
+	MessageType MessageType
+	Message     []byte
+	ID          string // optional
+}
+
+type Subscription struct {
+	dest  MessageSubscriber
+	types []MessageType
+}
+
+func NewMessageBus() MessageBus {
+	return MessageBus{
+		register:   make(chan Subscription, 10),
+		unregister: make(chan Subscription),
```

Comment:
Wondering if async register/unregister is a good idea - especially since they're different channels (i.e. no ordering guarantees between the `register` and `inbound` channels - so essentially you can never know when you're registered and guaranteed to receive a following message)

Reply:
removed in favor of direct registration